### PR TITLE
[core] Error check Redis get requests

### LINKED ISSUE #40412: [core] Silent failure when retrieving session key from Redis
ray.init may fail to retrieve a session key from Redis but continue executing, in which case it will generate a new session key, leading to two inconsistent keys (in Redis and passed in from the frontend).

### PR DESCRIPTION
**Pull request by vitsai**
It is possible for a Redis Get to silently fail in the current logic. Fix this to avoid weird downstream effects (such as session key inconsistency)

## Related issue number

https://github.com/ray-project/ray/issues/40412

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/store_client_kv.cc [hunk 1]
```diff
@@ -62,6 +62,7 @@ void StoreClientInternalKV::Get(
       table_name_,
       MakeKey(ns, key),
       [callback = std::move(callback)](auto status, auto result) {
+        RAY_CHECK(status.ok()) << "Failed to get from Redis with status: " << status;
```
Comment (iycheng) [inline src/ray/gcs/gcs_server/store_client_kv.cc:65]: Do you know why the redis client doesn't crash?

We probably shouldn't add it here. Try to put it into lower part.

Comment (vitsai) [reply src/ray/gcs/gcs_server/store_client_kv.cc]: Good point

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 2]
```diff
@@ -164,6 +164,8 @@ Status RedisStoreClient::AsyncGet(const std::string &table_name,
     if (!reply->IsNil()) {
       result = reply->ReadAsString();
     }
+    RAY_CHECK(!reply->IsError())
```
Comment (iycheng) [inline src/ray/gcs/store_client/redis_store_client.cc:167]: After reading the code, I think the right way to do is to update the callback right?

Comment (iycheng) [reply src/ray/gcs/store_client/redis_store_client.cc:167]: Should `callback(Status::OK(), std::move(result));`  become something else?

Comment (vitsai) [reply src/ray/gcs/store_client/redis_store_client.cc:167]: If the difference is between putting it in the callback in KVStore vs the callback in RedisStoreClient, I don't think it matters that much -- the lower callback starts first so crash will happen earlier without bubbling up.

Do you think we need to pass the status back up if we have a RAY_CHECK before it? It's guaranteed to be no error.

Comment (vitsai) [reply src/ray/gcs/store_client/redis_store_client.cc:167]: Synced offline. We can pass it back to the client instead of crashing GCS

**Comment by iycheng** [comment]
Synced offline. It seems that we have already crashed the program when receiving an error [here](https://github.com/ray-project/ray/blob/bc80271c5e20feb4263634a7334196483a90648f/src/ray/gcs/redis_context.cc#L45-L48) 

So the error might come from Connect [here](https://github.com/ray-project/ray/blob/master/python/ray/includes/global_state_accessor.pxd#L154C1-L158) 

Some simple test without Redis can verify this.

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -4624,4 +4624,5 @@ def get_session_key_from_storage(host, port, password, use_ssl, config, key):
     if result:
         return data
     else:
+        logger.warning("Could not retrieve session key from storage.")
```
Comment (iycheng) [inline python/ray/_raylet.pyx:4627]: Warning seems not right. Just logger.info.

**Review by fishbone — APPROVED** [review]
LG!

**Pull request merged**